sentry-ruby/lib/sentry/baggage.rb

```diff
@@ -0,0 +1,85 @@
+# frozen_string_literal: true
+
+require 'cgi'
+
+module Sentry
+  # A {https://www.w3.org/TR/baggage W3C Baggage Header} implementation.
+  class Baggage
+    SENTRY_PREFIX = 'sentry-'.freeze
+    SENTRY_PREFIX_REGEX = /^sentry-/.freeze
+
+    DSC_KEYS = %w(
+      trace_id
+      public_key
+      sample_rate
+      release
+      environment
+      transaction
+      user_id
+      user_segment
+    ).freeze
+
+    # @return [Boolean]
+    attr_reader :mutable
+
+    def initialize(sentry_items, third_party_items: '', mutable: true)
+      @sentry_items = sentry_items
+      @third_party_items = third_party_items
+      @mutable = mutable
+    end
+
+    # Creates a Baggage object from an incoming W3C Baggage header string.
+    #
+    # Sentry items are identified with the 'sentry-' prefix and stored in a hash.
+    # Third party items are stored verbatim in a separate string.
+    # The presence of a Sentry item makes the baggage object immutable.
+    #
+    # @param header [String] The incoming Baggage header string.
+    # @return [Baggage, nil]
+    def self.from_incoming_header(header)
+      return nil if header.nil? || header.empty?
+
+      sentry_items = {}
+      third_party_items = ''
+      mutable = true
+
+      header.split(',').map(&:strip).each do |item|
+        key, val = item.split('=')
+        next unless key && val
+
+        if key =~ SENTRY_PREFIX_REGEX
+          baggage_key = CGI.unescape(key.split('-')[1])
```

Comment:
Since key `"sentry-"` can also make it here, its `key.split('-')[1]` will be `nil`, which would cause `CGI.unescape` to raise argument error.